### ISSUE TITLE
Clang "-Wshadow" warnings

### ISSUE THREAD

**Issue by rmpowell77** [Microsoft/GSL#208]
When building with "-Wshadow" on clang-702.2.17.1:

```
GSL/span.h:510:43: error: declaration shadows a field of 'TypeListIndexer<TypeChain>' [-Werror,-Wshadow]
        TypeListIndexer(const TypeChain & obj) :obj(obj){}
                                          ^
GSL/span.h:509:27: note: previous declaration is here
        const TypeChain & obj;
                          ^
GSL/span.h:817:60: error: declaration shadows a field of 'bounds_iterator<IndexType>' [-Werror,-Wshadow]
    explicit bounds_iterator(const Bounds& bnd, value_type curr) noexcept
                                                           ^
GSL/span.h:1008:37: note: previous declaration is here
    std::remove_const_t<value_type> curr;
                                    ^
```

**Comment by gdr-at-ms** [comment]
This idiom is so common in C++ that a compiler that warns on it, makes the warning less useful by hiding it a sea of useless warnings.

**Comment by bdeeming** [comment]
The GSL is intended for extremely widespread use, and as such should support as many environments as possible. For example, we have this very warning enabled in our builds, and having a library that generated these warnings would be _highly_ annoying (and add just another hurdle to jump before we could adopt it).
This warning is very easy to fix, and has the potential to find real bugs, so I see no reason not to fix it and enable the warning flag in the main build.

**Comment by gdr-at-ms** [comment]
@bdeeming 
You may have misunderstood what I meant.
"Shadowing warning" is a useful one to have one's build set.
A bad implementation of "shadowing warning" is that useful.  For example, not all compilers implementing "shadowing warning" implement this bad version of it -- which actually has a special rule, specifically designed in the language to accommodate/support the practice.  Yes, I agree with you this particular warning is annoying, but only because the compiler decides to have an annoying/buggy implementation of something that is otherwise useful.

Just like we want GSL to be widely adopted, we are also careful not trying too hard to workaround every single bug in compilers.

**Comment by bdeeming** [comment]
@gdr-at-ms I do apologize, but I still do not understand what you mean. The example warning output shown above is exactly the output I would expect from the warning - it doesn't strike me as being the result of a bug in the compiler (here both the member variable `obj` and the function parameter `obj` are both available in the same scope, and therefore the function parameter over-shadows the member variable). We use GCC and it behaves in the same way as Clang does above. Are you able to point me to an example of a 'good' version of the warning?

Ps I don't intend to be annoying for the sake of it, I would just like to see the PR associated with this issue go through =)

**Comment by neilmacintosh** [comment]
@rmpowell77 Thanks for the report! @bdeeming I'm closing this issue because the warnings should be addressed by PR #216.
